compiler/rustc_span/src/span_encoding.rs

```diff
@@ -55,14 +62,19 @@ use rustc_data_structures::fx::FxIndexSet;
 ///   the code. No crates in `rustc-perf` need more than 15 bits for `ctxt`,
 ///   but larger crates might need more than 16 bits.
 ///
+/// In order to reliably use parented spans in incremental compilation,
+/// the dependency to the parent definition's span. This is performed
+/// using the callback `SPAN_TRACK` to access the query engine.
+///
 #[derive(Clone, Copy, Eq, PartialEq, Hash)]
 pub struct Span {
     base_or_index: u32,
     len_or_tag: u16,
     ctxt_or_zero: u16,
```

Comment:
The current span inlining strategy is a result of some noticeable amount of statistical and performance analysis, span interner is hit very rarely currently.
All that statistics need to be recalculated - we need to know which number is larger on average, `ctxt` or `parent`, and which should take more bits (maybe they'll even need to borrow some bits from `base_or_index`/`len_or_tag`).
Otherwise hitting the span interner more often is going to be pretty detrimental for performance.

Reply:
This is one of the most sensitive knobs. I did not really try to optimize the encoding, except by considering `ctxt` was less likely to be set than the rest. Do you have method recommendations to gather and analyse these stats?

Reply:
I'd like to understand where regressions in non-incremental mode come from, so for a start I'd suggest to
- never inline spans with `parent.is_some()`, so we don't get any regressions when the parent optimization is not done
- add a new `-Z` option enabling this optimization in `lower_span`
- then experiment with span inlining schemes later in a separate performance-focused PR